crates/vhost-device-gpio/src/backend.rs

```diff
@@ -214,33 +221,41 @@ fn start_backend(args: GpioArgs) -> Result<()> {
-            // These are run in an infinite loop to not require the daemon to be restarted once a
-            // guest exits.
-            //
-            // There isn't much value in complicating code here to return an error from the
-            // threads, and so the code uses unwrap() instead. The panic on a thread won't cause
-            // trouble to other threads/guests or the main() function and should be safe for the
-            // daemon.
-            match cfg {
-                GpioDeviceType::PhysicalDevice(device_num) => {
-                    let controller = PhysDevice::open(device_num).unwrap();
-                    start_device_backend(controller, socket.clone());
-                }
-                #[cfg(feature = "mock_gpio")]
-                GpioDeviceType::SimulatedDevice(num_gpios) => {
-                    let controller = MockGpioDevice::open(num_gpios).unwrap();
-                    start_device_backend(controller, socket.clone());
-                }
-                #[cfg(not(feature = "mock_gpio"))]
-                _ => {
-                    panic!("mock implementatoins aren't supported");
-                }
-            };
-        });
-
-        handles.push(handle);
+        let builder = std::thread::Builder::new();
+        let handle: Result<JoinHandle<Result<()>>> = builder
+            .spawn(move || loop {
+                // A separate thread is spawned for each socket and can connect to a separate guest.
+                // These are run in an infinite loop to not require the daemon to be restarted once a
+                // guest exits.
+                //
+                // There isn't much value in complicating code here to return an error from the
+                // threads, and so the code uses unwrap() instead. The panic on a thread won't cause
+                // trouble to other threads/guests or the main() function and should be safe for the
+                // daemon.
+                match cfg {
+                    GpioDeviceType::PhysicalDevice(device_num) => {
+                        let controller =
+                            PhysDevice::open(device_num).map_err(Error::CouldNotOpenDevice)?;
+                        start_device_backend(controller, socket.clone())?;
+                    }
+                    #[cfg(feature = "mock_gpio")]
+                    GpioDeviceType::SimulatedDevice(num_gpios) => {
+                        let controller =
+                            MockGpioDevice::open(num_gpios).map_err(Error::CouldNotOpenDevice)?;
+                        start_device_backend(controller, socket.clone())?;
+                    }
+                    #[cfg(not(feature = "mock_gpio"))]
+                    _ => {
+                        return Err(Error::DeviceUnimplemented);
+                    }
+                };
+            })
+            .map_err(|_| Error::FailedSpawningThread);
```

Comment:
Add `?` here.

Reply:
the switch back to the original approach renders this comment mute